experimental/traffic-portal/src/app/core/cache-groups/asns/table/asns-table.component.ts

```diff
@@ -0,0 +1,133 @@
+import { DecisionDialogComponent } from "src/app/shared/dialogs/decision-dialog/decision-dialog.component";
+import type { ContextMenuActionEvent, ContextMenuItem } from "src/app/shared/generic-table/generic-table.component";
+import { NavigationService } from "src/app/shared/navigation/navigation.service";
+
+/**
+ * AsnsTableComponent is the controller for the "Asns" table.
+ */
+@Component({
+	selector: "tp-asns",
+	styleUrls: ["./asns-table.component.scss"],
+	templateUrl: "./asns-table.component.html"
+})
+export class AsnsTableComponent implements OnInit {
+	/** List of asns */
+	public asns: Promise<Array<ResponseASN>>;
+
+	constructor(private readonly route: ActivatedRoute, private readonly headerSvc: NavigationService,
+		private readonly api: CacheGroupService, private readonly dialog: MatDialog, public readonly auth: CurrentUserService) {
+		this.fuzzySubject = new BehaviorSubject<string>("");
+		this.asns = this.api.getASNs();
+		this.headerSvc.headerTitle.next("ASNs");
+	}
+
+	/** Initializes table data, loading it from Traffic Ops. */
+	public ngOnInit(): void {
+		this.route.queryParamMap.subscribe(
+			m => {
+				const search = m.get("search");
+				if (search) {
+					this.fuzzControl.setValue(decodeURIComponent(search));
+					this.updateURL();
+				}
+			},
+			e => {
+				console.error("Failed to get query parameters:", e);
+			}
+		);
+	}
+
+	/** Definitions of the table's columns according to the ag-grid API */
+	public columnDefs = [
+		{
+			field: "asn",
+			headerName: "ASN"
+		},
+		{
+			field: "cachegroup",
+			headerName: "Cache Group",
+		},
+		{
+			field: "lastUpdated",
+			headerName: "Last Updated"
+		}
+	];
+
+	/** Definitions for the context menu items (which act on augmented asn data). */
+	public contextMenuItems: Array<ContextMenuItem<ResponseASN>> = [
+		{
+			href: (selectedRow: ResponseASN): string => `${selectedRow.id}`,
+			name: "Edit"
+		},
+		{
+			href: (selectedRow: ResponseASN): string => `${selectedRow.id}`,
+			name: "Open in New Tab",
+			newTab: true
+		},
+		{
+			action: "delete",
+			multiRow: false,
+			name: "Delete"
+		},
+		{
+			href: "/core/cache-groups",
+			name: "View Cache Group",
+			queryParams: (selectedRow: ResponseASN): Params => ({name: selectedRow.cachegroup}),
+		}
+	];
+
+	/** A subject that child components can subscribe to for access to the fuzzy search query text */
+	public fuzzySubject: BehaviorSubject<string>;
+
+	/** Form controller for the user search input. */
+	public fuzzControl = new FormControl<string>("");
+
+	/** Update the URL's 'search' query parameter for the user's search input. */
+	public updateURL(): void {
+		this.fuzzySubject.next(this.fuzzControl.value ?? "");
+	}
+
+	/**
+	 * Handles a context menu event.
+	 *
+	 * @param evt The action selected from the context menu.
+	 */
+	public async handleContextMenu(evt: ContextMenuActionEvent<ResponseASN>): Promise<void> {
+		const data = evt.data as ResponseASN;
```

Comment:
When you use `as`, it'd be nice to have a comment explaining why that's necessary/safe. The idea being that `as` should almost never be used (except in `as const`). In this case, we know that there are no multi-row actions, so this is a safe assumption to make - if you want to do that; personally I'd just check if it's an array and then throw an error because then you wouldn't need to use `as`.

Reply:
Ok, checking to see if it is an array.